override bool(), is_nonzero for real tensor tracing

pianpwk opened the pull request:
Fixes bool() and is_nonzero() calls for real tensor tracing, non-strict export

Differential Revision: D63482693

ezyang reviewed (changes requested):
Did overriding `__bool__` on fake tensor not work?

pianpwk commented:
> Did overriding `__bool__` on fake tensor not work?

@ezyang Noo, I guess in non-strict we're passing around FunctionalTensors

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -801,6 +801,10 @@ def can_handle_tensor(x: Tensor) -> bool:
 
+    if func is torch.ops.aten.is_nonzero.default:
+        with proxy_mode:
+            return (args[0] != 0).item()  # type: ignore[attr-defined]
```
This doesn't do anything anymore right?

pianpwk replied:
I guess this still enables export explicit is_nonzero calls in non-strict + real-tensor mode

ezyang commented on torch/_subclasses/fake_tensor.py:
```diff
@@ -1920,6 +1920,13 @@ def go(t: object, real_t: Tensor) -> None:
                         assert self.shape_env is not None
                         self.shape_env.set_unbacked_var_to_val(t.node.expr, real_t)
+                    elif (
+                        isinstance(s := t.node.expr, sympy.Eq)
```
sigh, indicator variables, the gift that keeps giving (no action needed here)

pianpwk committed:
override bool(), is_nonzero for real tensor tracing (#136788)

Summary:
Fixes bool() and is_nonzero() calls for real tensor tracing, non-strict export


Test Plan: test_export

Reviewed By: ezyang

Differential Revision: D63482693
